# Project: Data Wrangling and Analysis on WeRateDogs' Twitter page data.

# <u>Gathering

#### Importing the necessarily libraries 

In [1]:
import pandas as pd
import requests
import os
import json

#### Importing the first dataset

In [2]:
df_archive = pd.read_csv(r'C:\Users\PC\Documents\GitHub\ALX-T\Project 02 - Data Wrangling\Datasets\twitter-archive-enhanced.csv')

#### Downloading the second dataset programmatically using the requests library

In [3]:
folder_name = 'Datasets'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [4]:
with open(os.path.join(folder_name, url.split('/')[-1]), mode = 'wb') as file:
    file.write(response.content)

In [5]:
df_images = pd.read_csv(r'C:\Users\PC\Documents\GitHub\ALX-T\Project 02 - Data Wrangling\Datasets\image-predictions.tsv', sep='\t')

#### Reading the third dataset into a dataframe from the tweet-json.txt file 

I had issues getting access to the twitter API, so i had to get the additional twitter data via Udacity's servers.

In [6]:
twitter = []
with open('tweet-json.txt', 'r') as Json_file:
    for line in Json_file:
        tweet = (json.loads(line))
        tweet_id = tweet['id']
        retweet_count = tweet['retweet_count']
        favorite_count = tweet['favorite_count']
        create_date = tweet['created_at']
        twitter.append({'retweet_count' : retweet_count,
                  'favorite_count' : favorite_count,
                  'create_date' : create_date,
                  'tweet_id' : tweet_id})

In [7]:
df_twitter = pd.DataFrame(twitter, columns = ['tweet_id', 'retweet_count', 'favorite_count', 'create_date'])

# <u>Assessing


This may be done visually or programmatically. First, we will attempt doing it visually by just printing the dataset and trying to identify any errors.

### Visual Assessment

In [8]:
df_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [9]:
df_images

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [10]:
df_twitter

tweet_id  retweet_count  favorite_count  \
0     892420643555336193           8853           39467   
1     892177421306343426           6514           33819   
2     891815181378084864           4328           25461   
3     891689557279858688           8964           42908   
4     891327558926688256           9774           41048   
...                  ...            ...             ...   
2349  666049248165822465             41             111   
2350  666044226329800704            147             311   
2351  666033412701032449             47             128   
2352  666029285002620928             48             132   
2353  666020888022790149            532            2535   

                         create_date  
0     Tue Aug 01 16:23:56 +0000 2017  
1     Tue Aug 01 00:17:27 +0000 2017  
2     Mon Jul 31 00:18:03 +0000 2017  
3     Sun Jul 30 15:58:51 +0000 2017  
4     Sat Jul 29 16:00:24 +0000 2017  
...                              ...  
2349  Mon Nov 16 00:24:50 +0000 2015  
2350  Mon Nov 16 00:04:52 +0000 2015  
2351  Sun Nov 15 23:21:54 +0000 2015  
2352  Sun Nov 15 23:05:30 +0000 2015  
2353  Sun Nov 15 22:32:08 +0000 2015  

[2354 rows x 4 columns]

This data set is large, and as a result most of the data has been collapsed. However, there are still some obvious quality and tidiness issues that may be observed.

### Programmatic Assessment

This is done with the use of python code, and some python libraries.

In order to save time, and avoid code repetition, we will create a function which will return basic information on our datasets.

In [11]:
def info(df):
    print(df.info())
    print(df.shape)

#### df_archive

In [12]:
info(df_archive)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [15]:
df_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [29]:
df_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1409  699060279947165696                    NaN                  NaN   
2192  668932921458302977                    NaN                  NaN   
817   770772759874076672                    NaN                  NaN   
2261  667549055577362432                    NaN                  NaN   
100   872820683541237760                    NaN                  NaN   

                      timestamp  \
1409  2016-02-15 02:38:53 +0000   
2192  2015-11-23 23:23:31 +0000   
817   2016-08-30 23:58:40 +0000   
2261  2015-11-20 03:44:31 +0000   
100   2017-06-08 14:20:41 +0000   

                                                 source  \
1409  <a href="http://vine.co" rel="nofollow">Vine -...   
2192  <a href="http://twitter.com/download/iphone" r...   
817   <a href="http://twitter.com/download/iphone" r...   
2261  <a href="http://twitter.com" rel="nofollow">Tw...   
100   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1409  This is Yukon. He pukes rainbows. 12/10 magica...                  NaN   
2192  This is Herald. He likes to swing. Subtle tong...                  NaN   
817   This is Sebastian. He's super h*ckin fluffy. T...                  NaN   
2261  Never seen dog like this. Breathes heavy. Tilt...                  NaN   
100   Here are my favorite #dogsatpollingstations \n...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1409                       NaN                        NaN   
2192                       NaN                        NaN   
817                        NaN                        NaN   
2261                       NaN                        NaN   
100                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1409                      https://vine.co/v/inlmMHxtqDD                12   
2192  https://twitter.com/dog_rates/status/668932921...                 9   
817   https://twitter.com/dog_rates/status/770772759...                11   
2261  https://twitter.com/dog_rates/status/667549055...                 1   
100   https://twitter.com/dog_rates/status/872820683...                13   

      rating_denominator       name doggo floofer pupper puppo  
1409                  10      Yukon  None    None   None  None  
2192                  10     Herald  None    None   None  None  
817                   10  Sebastian  None    None   None  None  
2261                  10       None  None    None   None  None  
100                   10       None  None    None   None  None

Checking appropriate columns for duplicate entries

In [22]:
df_archive[df_archive['tweet_id'].duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [24]:
df_archive[df_archive['text'].duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [25]:
df_archive[df_archive['expanded_urls'].duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
75    878281511006478336                    NaN                  NaN   
76    878057613040115712                    NaN                  NaN   
98    873213775632977920                    NaN                  NaN   
...                  ...                    ...                  ...   
2286  667182792070062081                    NaN                  NaN   
2293  667152164079423490                    NaN                  NaN   
2294  667138269671505920                    NaN                  NaN   
2298  667070482143944705           6.670655e+17         4.196984e+09   
2338  666104133288665088                    NaN                  NaN   

                      timestamp  \
55    2017-07-02 21:58:53 +0000   
64    2017-06-27 12:14:36 +0000   
75    2017-06-23 16:00:04 +0000   
76    2017-06-23 01:10:23 +0000   
98    2017-06-09 16:22:42 +0000   
...                         ...   
2286  2015-11-19 03:29:07 +0000   
2293  2015-11-19 01:27:25 +0000   
2294  2015-11-19 00:32:12 +0000   
2298  2015-11-18 20:02:51 +0000   
2338  2015-11-16 04:02:55 +0000   

                                                 source  \
55    <a href="http://twitter.com/download/iphone" r...   
64    <a href="http://twitter.com/download/iphone" r...   
75    <a href="http://twitter.com/download/iphone" r...   
76    <a href="http://twitter.com/download/iphone" r...   
98    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2286  <a href="http://twitter.com/download/iphone" r...   
2293  <a href="http://twitter.com/download/iphone" r...   
2294  <a href="http://twitter.com/download/iphone" r...   
2298  <a href="http://twitter.com/download/iphone" r...   
2338  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
55    @roushfenway These are good dogs but 17/10 is ...                  NaN   
64                      @RealKentMurphy 14/10 confirmed                  NaN   
75    Meet Shadow. In an attempt to reach maximum zo...                  NaN   
76    This is Emmy. She was adopted today. Massive r...                  NaN   
98    This is Sierra. She's one precious pupper. Abs...                  NaN   
...                                                 ...                  ...   
2286  This is Timison. He just told an awful joke bu...                  NaN   
2293  This is Pipsy. He is a fluffball. Enjoys trave...                  NaN   
2294  Extremely intelligent dog here. Has learned to...                  NaN   
2298  After much debate this dog is being upgraded t...                  NaN   
2338  Not familiar with this breed. No tail (weird)....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
55                         NaN                        NaN   
64                         NaN                        NaN   
75                         NaN                        NaN   
76                         NaN                        NaN   
98                         NaN                        NaN   
...                        ...                        ...   
2286                       NaN                        NaN   
2293                       NaN                        NaN   
2294                       NaN                        NaN   
2298                       NaN                        NaN   
2338                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
55                                                  NaN                17   
64                                                  NaN                14   
75    https://www.gofundme.com/3yd6y1c,https://twitt...                13   
76    https://twitt

Checking appropriate columns for unique values

In [40]:
df_archive.name.value_counts()

None          745
a              55
Charlie        12
Cooper         11
Lucy           11
             ... 
Dex             1
Ace             1
Tayzie          1
Grizzie         1
Christoper      1
Name: name, Length: 957, dtype: int64

In [35]:
df_archive.loc[df_archive['name'] == 'None']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
5     2017-07-29 00:08:17 +0000   
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2342  2015-11-16 02:38:37 +0000   
2343  2015-11-16 01:59:36 +0000   
2344  2015-11-16 01:52:02 +0000   
2351  2015-11-16 00:24:50 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
5     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2342  <a href="http://twitter.com/download/iphone" r...   
2343  <a href="http://twitter.com/download/iphone" r...   
2344  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5     Here we have a majestic great white breaching ...                  NaN   
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2342  Here we have a well-established sunblockerspan...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
2344  Here we have a northern speckled Rhododendron....                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5                          NaN                        NaN   
7                          NaN                        NaN   
12                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2342                       NaN                        NaN   
2343                       NaN                        NaN   
2344                       NaN                        NaN   
2351                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5     https://twitter.com/dog_rates/status/891087950...                13   
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
24    https://twitt

In [52]:
#finding columns that have a retweeted status id
df_archive[df_archive['retweeted_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
...                  ...                    ...                  ...   
1023  746521445350707200                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
2259  667550904950915073                    NaN                  NaN   
2260  667550882905632768                    NaN                  NaN   

                      timestamp  \
19    2017-07-21 01:02:36 +0000   
32    2017-07-15 02:45:48 +0000   
36    2017-07-13 01:35:06 +0000   
68    2017-06-26 00:13:58 +0000   
73    2017-06-24 00:09:53 +0000   
...                         ...   
1023  2016-06-25 01:52:36 +0000   
1043  2016-06-17 16:01:16 +0000   
1242  2016-03-21 19:31:59 +0000   
2259  2015-11-20 03:51:52 +0000   
2260  2015-11-20 03:51:47 +0000   

                                                 source  \
19    <a href="http://twitter.com/download/iphone" r...   
32    <a href="http://twitter.com/download/iphone" r...   
36    <a href="http://twitter.com/download/iphone" r...   
68    <a href="http://twitter.com/download/iphone" r...   
73    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1023  <a href="http://twitter.com/download/iphone" r...   
1043  <a href="http://twitter.com/download/iphone" r...   
1242  <a href="http://twitter.com/download/iphone" r...   
2259  <a href="http://twitter.com" rel="nofollow">Tw...   
2260  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
19    RT @dog_rates: This is Canela. She attempted s...         8.874740e+17   
32    RT @Athletics: 12/10 #BATP https://t.co/WxwJmv...         8.860537e+17   
36    RT @dog_rates: This is Lilly. She just paralle...         8.305833e+17   
68    RT @dog_rates: This is Emmy. She was adopted t...         8.780576e+17   
73    RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
...                                                 ...                  ...   
1023  RT @dog_rates: This is Shaggy. He knows exactl...         6.678667e+17   
1043  RT @dog_rates: Extremely intelligent dog here....         6.671383e+17   
1242  RT @twitter: @dog_rates Awesome Tweet! 12/10. ...         7.119983e+17   
2259  RT @dogratingrating: Exceptional talent. Origi...         6.675487e+17   
2260  RT @dogratingrating: Unoriginal idea. Blatant ...         6.675484e+17   

      retweeted_status_user_id retweeted_status_timestamp  \
19                4.196984e+09  2017-07-19 00:47:34 +0000   
32                1.960740e+07  2017-07-15 02:44:07 +0000   
36                4.196984e+09  2017-02-12 01:04:29 +0000   
68                4.196984e+09  2017-06-23 01:10:23 +0000   
73                4.196984e+09  2017-06-23 16:00:04 +0000   
...                        ...                        ...   
1023              4.196984e+09  2015-11-21 00:46:50 +0000   
1043              4.196984e+09  2015-11-19 00:32:12 +0000   
1242              7.832140e+05  2016-03-21 19:29:52 +0000   
2259              4.296832e+09  2015-11-20 03:43:06 +0000   
2260              4.296832e+09  2015-11-20 03:41:59 +0000   

                                          expanded_urls  rating_numerator  \
19    https://twitter.com/dog_rates/status/887473957...                13   
32    https://twitter.com/dog_rates/status/886053434...                12   
36    https://twitter.com/dog_rates/status/830583320...                13   
68    https://twitt

#### df_images

In [16]:
info(df_images)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB
None
(2075, 12)


In [17]:
df_images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [41]:
df_images.sample(5)

tweet_id                                          jpg_url  \
1658  810984652412424192  https://pbs.twimg.com/media/C0EyPZbXAAAceSc.jpg   
1460  778039087836069888  https://pbs.twimg.com/media/CswmaHmWAAAbdY9.jpg   
1841  838083903487373313  https://pbs.twimg.com/media/C6F42cGUYAAIKsX.jpg   
1544  791672322847637504  https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1357  760641137271070720  https://pbs.twimg.com/media/Co5XExUWgAAL5L_.jpg   

      img_num                p1   p1_conf  p1_dog                  p2  \
1658        1  golden_retriever  0.871342    True     Tibetan_mastiff   
1460        2   German_shepherd  0.717776    True            malinois   
1841        2              chow  0.800975    True           seat_belt   
1544        1  golden_retriever  0.705092    True  Labrador_retriever   
1357        1           axolotl  0.132695   False        killer_whale   

       p2_conf  p2_dog                  p3   p3_conf  p3_dog  
1658  0.036708    True  Labrador_retriever  0.025823    True  
1460  0.111175    True  Norwegian_elkhound  0.058802    True  
1841  0.164133   False          Pomeranian  0.017981    True  
1544  0.219721    True              kuvasz  0.015965    True  
1357  0.131113   False            sea_lion  0.069652   False

Checking appropriate columns for duplicate values.

In [46]:
df_images[df_images['jpg_url'].duplicated()]

tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
...                  ...                                                ...   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
1992  873697596434513921    https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   

      img_num                         p1   p1_conf  p1_dog  \
1297        1                    upright  0.303415   False   
1315        1                        pug  0.272205    True   
1333        2               cash_machine  0.802333   False   
1345        1              Irish_terrier  0.254856    True   
1349        1         Labrador_retriever  0.967397    True   
...       ...                        ...       ...     ...   
1903        1  Staffordshire_bullterrier  0.757547    True   
1944        2              Arabian_camel  0.366248   False   
1992        1                     laptop  0.153718   False   
2041        1         Labrador_retriever  0.908703    True   
2055        2                   Pembroke  0.809197    True   

                                  p2   p2_conf  p2_dog  \
1297                golden_retriever  0.181351    True   
1315                    bull_mastiff  0.251530    True   
1333                      schipperke  0.045519    True   
1345                          briard  0.227716    True   
1349                golden_retriever  0.016641    True   
...                              ...       ...     ...   
1903  American_Staffordshire_terrier  0.149950    True   
1944                     house_finch  0.209852   False   
1992                  French_bulldog  0.099984    True   
2041                       seat_belt  0.057091   False   
2055             Rhodesian_ridgeback  0.054950    True   

                               p3   p3_conf  p3_dog  
1297             Brittany_spaniel  0.162084    True  
1315                   bath_towel  0.116806   False  
1333              German_shepherd  0.023353    True  
1345  soft-coated_wheaten_terrier  0.223263    True  
1349                     ice_bear  0.014858   False  
...                           ...       ...     ...  
1903     Chesapeake_Bay_retriever  0.047523    True  
1944               cocker_spaniel  0.046403    True  
1992                      printer  0.077130   False  
2041                          pug  0.011933    True  
2055                       beagle  0.038915    True  

[66 rows x 12 columns]

In [44]:
df_images.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
pillow                  1
carousel                1
bald_eagle              1
lorikeet                1
orange                  1
Name: p1, Length: 378, dtype: int64

#### df_twitter

In [14]:
info(df_twitter)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2354 non-null   int64 
 1   retweet_count   2354 non-null   int64 
 2   favorite_count  2354 non-null   int64 
 3   create_date     2354 non-null   object
dtypes: int64(3), object(1)
memory usage: 73.7+ KB
None
(2354, 4)


In [19]:
df_twitter.describe()

tweet_id  retweet_count  favorite_count
count  2.354000e+03    2354.000000     2354.000000
mean   7.426978e+17    3164.797366     8080.968564
std    6.852812e+16    5284.770364    11814.771334
min    6.660209e+17       0.000000        0.000000
25%    6.783975e+17     624.500000     1415.000000
50%    7.194596e+17    1473.500000     3603.500000
75%    7.993058e+17    3652.000000    10122.250000
max    8.924206e+17   79515.000000   132810.000000

### Issues

#### df_archive
Quality Issues
- Missing Data: in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls
- Lowercase dog name
- Erroneous datatypes: in_reply_to_status_id, in_reply_to_user_id, timestamp, 
- Duplicate Entries: expanded_urls
- Erroneous entries: null values in names are entered as "none" instead of "NaN"
- Retweeted rows 

Tidiness
- Separating the text column into name and breed 

#### df_images
Quality Issues
- Lowercase dog name
- Missing rows: we should have 2356
- Duplicate Entries: jpg_url

Tidiness Issues
- The p1, p2 and p3 columns should form one column(dog breed), which would consist of the prediction with the highest p_conf

#### df_twitter
Quality Issues
- Create_date is object(string) instead of datetime datatype
- Missing rows we should have 2356